physics/discrete_trajectory_body.hpp

```diff
@@ -158,6 +160,44 @@ void DiscreteTrajectory<Frame>::Append(
   CHECK(--timeline_.end() == it)
       << "Append out of order at " << time << ", last time is "
       << (--timeline_.end())->first;
+  if (downsampling_.has_value()) {
+    if (timeline_.size() == 1) {
+      downsampling_->start_of_dense_timeline = timeline_.begin();
+      downsampling_->dense_intervals = 0;
+    } else {
+      CheckNoForksBefore(last().time());
+      ++downsampling_->dense_intervals;
+      if (downsampling_->dense_intervals ==
+          downsampling_->max_dense_intervals) {
+        std::vector<TimelineConstIterator> dense_iterators;
+        dense_iterators.reserve(downsampling_->max_dense_intervals + 1);
+        for (TimelineConstIterator it = downsampling_->start_of_dense_timeline;
+             it != timeline_.end();
+             ++it) {
+          dense_iterators.push_back(it);
+        }
+        auto right_endpoints = FitHermiteSpline<Instant, Position<Frame>>(
+            dense_iterators,
+            [](auto&& it) -> auto&& { return it->first; },
+            [](auto&& it) -> auto&& { return it->second.position(); },
+            [](auto&& it) -> auto&& { return it->second.velocity(); },
+            downsampling_->tolerance);
+        if (right_endpoints.empty()) {
+          right_endpoints.push_back(dense_iterators.end() - 1);
+        }
+        TimelineConstIterator left = downsampling_->start_of_dense_timeline;
+        for (const auto& it_in_dense_iterators : right_endpoints) {
```

Comment:
I find it odd to call `it_in_dense_iterators` an iterator which is not in `dense_iterators`.

Reply:
It is.